### ISSUE TITLE
Single and Double often give too long or too short buzzes

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/quake#27]
For https://github.com/phetsims/qa/issues/842. This was seen on a tablet that does not have haptics, so was instead noticed with sound enabled.
On the first screen, I noticed that between uses, the single and double button would show different lengths, sometimes seeming too long or short. With double in particular I've noticed instances where one buzz was longer or shorter than the other, or the two buzzes merge together. This seemed to happen most often after changing the length of a buzz, but could also be persistent, as shown in the video.

https://user-images.githubusercontent.com/41024075/195149386-d35f5dfa-0fc3-4811-b31f-8c5cdabc5f68.MOV

**Comment by KatieWoe** [comment]
Not everything is shown in the video, but things seem to be all over the place in general.

**Comment by KatieWoe** [comment]
Short Buzzes on the Buzzes screen can also be a bit short or long from press to press.

**Comment by KatieWoe** [comment]
Patterns on the Patterns screen also sometimes run together.

**Comment by jbphet** [comment]
For future reference, the device used in the video above is a Samsung Galaxy Tab A7 Lite, model SM-T220, S/N R9JR702MX3Y, that generally resides in the QA office.

This problem does not occur on the A7 that I used during the development of the app, nor does it occur when running the browser-based version of the app, but I just grabbed the A7 Lite out of the QA office and am able to easily duplicate the problematic behavior.

**Comment by jbphet** [comment]
I reviewed the sound generation code pretty carefully and made several modifications that I thought might improve the situation, but it didn't resolve the problem.  I committed the changes anyway, since they seem like an improvement to the code in terms of maintainability and general performance.

I then tried logging the timing of the calls to start and stop each segment of the sound patterns using `window.performance.now()`, and learned that as far as the code is concerned, it is starting and stopping the sounds at the correct time intervals.  This makes me think that the WebAudio rendering thread is where the problem lies, and that it isn't getting enough of the processor time to execute the gain change operations with accurate timing.  That isn't good news, since I don't know of any way to improve this.

For more information on the WebAudio https://www.w3.org/TR/webaudio/#processing-model-background.

**Comment by jbphet** [comment]
During the testing described above, at one point - and only one point - I got a warning on the Android Studio console that said:

```
2022-10-17 14:21:48.758 18167-18761/edu.colorado.phet.hapticsPlayground W/PluginManager: THREAD WARNING: exec() call to NativeVibration.vibrate blocked the main thread for 415ms. Plugin should use CordovaInterface.getThreadPool().
```

This got me thinking that perhaps the calls to the native vibration plugin are causing the timing problems, so I tried re-writing the `execute` call in `NativeVibration.java` to use a thread.  Alas, this didn't fix the problem either, so I backed it out.  The code is in the "details" section below in case I ever want to return to it.  After that experiment, I tried completely commenting out the call to the haptic vibration methods and made a "sound only" version of the app, but that still has the inconsistent timing problem on the Tab A7 Lite, so the calls to the native vibration plugin are clearly *not* at the root of the problem.

<details>

```java
    /**
     * Executes the request and returns PluginResult.
     *
     * @param action          - the action to execute
     * @param args            - JSONArry of arguments for the plugin
     * @param callbackContext - the callback id used when calling back into JavaScript
     * @return true if the action was valid, false if not
     */
    public boolean execute( String action, JSONArray args, CallbackContext callbackContext ) throws JSONException {
        Vibrator vibrator = this.vibrator;
        Log.i( "NativeVibration", "received action request: " + action );
        if ( "vibrate".equals( action ) ) {
            cordova.getThreadPool().execute( new Runnable() {
                public void run() {
                    try {
                        Log.i( "NativeVibration", "pattern spec: " + args.getJSONArray( 0 ).toString() );
                        boolean repeat = args.getBoolean( 1 );
                        Log.i( "NativeVibration", "repeat: " + String.valueOf( repeat ) );
                        int repeatIndex = repeat ? 0 : -1;
                        JSONArray patternSpec = args.getJSONArray( 0 );
                        long[] durations = new long[patternSpec.length()];
                        int[] intensities = new int[patternSpec.length()];
                        for ( int i = 0; i < patternSpec.length(); i++ ) {
                            JSONObject vibrationSpec = patternSpec.getJSONObject( i );
                            durations[i] = Math.round( vibrationSpec.getDouble( "duration" ) * 1000 );
                            int adjustedIntensity = (int) Math.round( vibrationSpec.getDouble( "intensity" ) * 255 );
                            intensities[i] = adjustedIntensity;
                        }
                        vibrator.vibrate( VibrationEffect.createWaveform( durations, intensities, repeatIndex ) );
                        JSONObject r = new JSONObject();
                        callbackContext.success( r );
                    }
                    catch( Exception e ) {
                        Log.i( "NativeVibration", "Exception thrown while trying to vibrate: " + e.toString() );
                        callbackContext.error( e.toString() );
                    }

                }
            } );

        }
        else if ( "cancel".equals( action ) ) {
            cordova.getThreadPool().execute( new Runnable() {
                public void run() {
                    vibrator.cancel();
                    JSONObject r = new JSONObject();
                    callbackContext.success( r );
                }
            } );
        }
        else {
            return false;
        }
        return true;
    }
```

</details>

**Comment by jbphet** [comment]
@stemilymill - I believe you are the only one on the QA team who tested using a device that actually supports haptics.  Did you experience any of the timing problems that @KatieWoe describes above?

**Comment by jbphet** [comment]
I may be able to fix this using a plugin that takes advantage of native sound.  There appears to be a well-supported plugin for this called cordova-plugin-nativeaudio, see https://www.npmjs.com/package/cordova-plugin-nativeaudio.  I talked with @emily-phet about whether I should spend the time to try this, and we decided that if I can fix the problem in roughly 3 hrs time I should go for it.  If it looks like it would take longer, I should bail and we will just release it with this behavior and note it in the product description.

**Comment by stemilymill** [comment]
@jbphet i have not experienced any of these timing issues, with or without sound turned on.

**Comment by jbphet** [comment]
I tried using two different native sound plugins, `cordova-plugin-nativeaudio` and `cordova-plugin-media`.  Of the two, I was able to get way further using the 2nd one.  I tried replacing the current Web-Audio-based sound generation with the native implementation, but the timing was only marginally better, and it had the undesirable effect of making the audio very clicky on the browser and other Android devices.  I have committed the code to a branch, see https://github.com/phetsims/quake/issues/28, in case we need to revisit this in the future, but I'm not going to move this into the mainline code.

Since I wasn't able to find a decent solution in the allotted time, I'm going to move forward with the publication process, and will note in the description of the app that the sound timing isn't great on non-haptics-supporting devices.  I wouldn't expect many people to use it that way anyways.  I'll give @KatieWoe the Samsung S7 for her to test on.
